src/OpenSage.Game/Logic/Orders/OrderProcessor.cs

```diff
@@ -93,7 +93,25 @@ public void Process(IEnumerable<Order> orders)
 
                             foreach (var unit in player.SelectedUnits)
                             {
-                                unit.ProductionUpdate.QueueProduction(objectDefinition);
+                                // Only units that can produce stuff should produce it
+                                unit?.ProductionUpdate.QueueProduction(objectDefinition);
+                            }
+                        }
+                        break;
+
+                    case OrderType.CancelUnit:
```

Comment:
Now cancelling a unit grants money back and is also handled through the correct order